File: crates/simcity_sim/src/game/buildings/population.rs
use bevy::prelude::*;

use crate::game::sim::City;

use super::components::Building;

/// Update city population from building occupancy.
/// Population is the sum of occupancy_residents from all operational buildings.
pub fn update_city_population(mut city: ResMut<City>, q_buildings: Query<&Building>) {
    let total_population: u32 = q_buildings
        .iter()
        .filter(|b| matches!(b.phase, super::components::BuildingPhase::Operational))
        .map(|b| b.occupancy_residents as u32)
        .sum();

    city.population = total_population;
}

File: packages/sim/src/buildings/population.ts
// Port of crates/simcity_sim/src/game/buildings/population.rs: the city's population is the residents of its operational buildings.
import type { World } from '../world';
import { isOperational } from './building';

export function updateCityPopulation(w: World): void {
  let population = 0;
  for (const b of w.buildings.all()) if (isOperational(b)) population += b.occupancyResidents;
  w.city.population = population;
}
